yarn/history/src/main/scala/org/apache/spark/deploy/history/yarn/YarnHistoryService.scala

```diff
@@ -0,0 +1,981 @@
+   * That is: not unique enough to be used as the entity ID
+   * @param appId application ID
+   * @param attemptId attempt ID
+   */
+  private def setContextAppAndAttemptInfo(appId: Option[String],
+      attemptId: Option[String]): Unit = {
+    logDebug(s"Setting application ID to $appId; attempt ID to $attemptId")
+    sparkApplicationId = appId
+    sparkApplicationAttemptId = attemptId
+  }
+
+  /**
+   * Add the listener if it is not disabled.
+   * This is accessible in the same package purely for testing
+   * @return true if the register was enabled
+   */
+  private [yarn] def registerListener: Boolean = {
+    assert(sparkContext != null, "Null context")
+    if (sparkContext.conf.getBoolean(REGISTER_LISTENER, true)) {
+      logDebug("Registering listener to spark context")
+      val l = new YarnEventListener(sparkContext, this)
+      listener = Some(l)
+      sparkContext.listenerBus.addListener(l)
+      true
+    } else {
+      false
+    }
+  }
+
+  /**
+   * Queue an action, or if the service's `stopped` flag
+   * is set, discard it
+   * @param action action to process
+   * @return true if the event was queued
+   */
+  def enqueue(action: QueuedAction): Boolean = {
+    if (!stopped.get) {
+      innerEnqueue(action)
+      true
+    } else {
+      if (timelineServiceEnabled) {
+        // if the timeline service was ever enabled, log the fact the event
+        // is being discarded. Don't do this if it was not, as it will
+        // only make the logs noisy.
+        logInfo(s"History service stopped; ignoring queued event : $action")
+      }
+      false
+    }
+  }
+
+  /**
+   * Inner operation to queue the event. This does not check for service state
+   * @param event event to queue
+   */
+  private def innerEnqueue(event: QueuedAction) = {
+    eventsQueued.incrementAndGet
+    logDebug(s"Enqueue $event")
+    actionQueue.add(event)
+  }
+
+  /**
+   * Stop the service; this triggers flushing the queue and, if not already processed,
+   * a pushing out of an application end event.
+   *
+   * This operation will block for up to `maxTimeToWaitOnShutdown` milliseconds
+   * to await the asynchronous action queue completing.
+   */
+  override protected def serviceStop {
+    // if the queue is live
+    if (!stopped.get) {
+
+      if (!appEndEventProcessed.get()) {
+        // push out an application stop event if none has been received
+        logDebug("Generating a SparkListenerApplicationEnd during service stop()")
+        val current = now()
+        innerEnqueue(new HandleSparkEvent(SparkListenerApplicationEnd(current), current))
+      }
+
+      // flush
+      logInfo(s"Shutting down: pushing out ${actionQueue.size} events")
+      innerEnqueue(FlushTimelineEvents())
+
+      // stop operation
+      postThreadActive synchronized {
+        innerEnqueue(StopQueue())
+        // now await that marker flag
+        postThreadActive.wait(maxTimeToWaitOnShutdown)
```

Comment:
minor, but this can miss a `notifyAll` if the thread exits for any reason that is not the `StopQueue` message. If you check the value of `postThreadActive` before waiting, you could avoid that.